File: pkg/idgen/snowflake.go
// Package idgen provides a lightweight snowflake-like ID generator.
// bfs uses gosnowflake; jfs embeds a local worker for zero external deps.
package idgen

import (
	"errors"
	"sync"
	"time"
)

const (
	epoch          int64 = 1609459200000 // 2021-01-01 UTC, ms
	workerBits           = 10
	sequenceBits         = 12
	maxWorkerID          = -1 ^ (-1 << workerBits)
	maxSequence          = -1 ^ (-1 << sequenceBits)
	workerShift          = sequenceBits
	timestampShift       = sequenceBits + workerBits
)

// Generator produces unique int64 keys.
type Generator struct {
	mu        sync.Mutex
	workerID  int64
	sequence  int64
	lastStamp int64
}

// New creates a generator bound to workerID (0..1023).
func New(workerID int64) (*Generator, error) {
	if workerID < 0 || workerID > maxWorkerID {
		return nil, errors.New("idgen: worker id out of range")
	}
	return &Generator{workerID: workerID}, nil
}

// Next returns a unique int64 key.
func (g *Generator) Next() (int64, error) {
	g.mu.Lock()
	defer g.mu.Unlock()

	now := time.Now().UnixMilli()
	if now < g.lastStamp {
		return 0, errors.New("idgen: clock moved backwards")
	}
	if now == g.lastStamp {
		g.sequence = (g.sequence + 1) & maxSequence
		if g.sequence == 0 {
			for now <= g.lastStamp {
				now = time.Now().UnixMilli()
			}
		}
	} else {
		g.sequence = 0
	}
	g.lastStamp = now
	id := ((now - epoch) << timestampShift) | (g.workerID << workerShift) | g.sequence
	return id, nil
}

// NextCookie returns a non-zero random-ish cookie derived from the key.
func NextCookie(key int64) int32 {
	c := int32(key ^ (key >> 32))
	if c == 0 {
		return 1
	}
	if c < 0 {
		return -c
	}
	return c
}

File: jfs-common/src/main/java/com/levis9527/jfs/idgen/Snowflake.java
package com.levis9527.jfs.idgen;

/**
 * Lightweight snowflake-like ID generator (bfs uses gosnowflake).
 */
public final class Snowflake {
    private static final long EPOCH = 1609459200000L; // 2021-01-01 UTC
    private static final long WORKER_BITS = 10L;
    private static final long SEQUENCE_BITS = 12L;
    private static final long MAX_WORKER = ~(-1L << WORKER_BITS);
    private static final long MAX_SEQUENCE = ~(-1L << SEQUENCE_BITS);
    private static final long WORKER_SHIFT = SEQUENCE_BITS;
    private static final long TIMESTAMP_SHIFT = SEQUENCE_BITS + WORKER_BITS;

    private final long workerId;
    private long sequence;
    private long lastStamp = -1L;

    public Snowflake(long workerId) {
        if (workerId < 0 || workerId > MAX_WORKER) {
            throw new IllegalArgumentException("worker id out of range: " + workerId);
        }
        this.workerId = workerId;
    }

    public synchronized long nextId() {
        long now = System.currentTimeMillis();
        if (now < lastStamp) {
            throw new IllegalStateException("clock moved backwards");
        }
        if (now == lastStamp) {
            sequence = (sequence + 1) & MAX_SEQUENCE;
            if (sequence == 0) {
                while (now <= lastStamp) {
                    now = System.currentTimeMillis();
                }
            }
        } else {
            sequence = 0;
        }
        lastStamp = now;
        return ((now - EPOCH) << TIMESTAMP_SHIFT) | (workerId << WORKER_SHIFT) | sequence;
    }

    /** Non-zero cookie derived from key (mitigate brute-force store access). */
    public static int nextCookie(long key) {
        int c = (int) (key ^ (key >>> 32));
        if (c == 0) {
            return 1;
        }
        return c < 0 ? -c : c;
    }
}
